src/validation/ValidationContext.js

```diff
@@ -129,14 +130,21 @@ export default class ValidationContext {
     return spreads;
   }
 
+  /*
+   * Finds all fragments referenced via the definition, recursively.
+   *
+   * NOTE: if experimentalFragmentVariables are being used, it excludes all
+   * fragments with their own variable definitions: these are considered their
+   * own "root" executable definition.
+   */
   getRecursivelyReferencedFragments(
```

Comment:
I'm very concern about changes in this function since it's called `getRecursivelyReferencedFragments` and is part of public API.
+ it's break `NoUnusedFragments`

Alternative solution would be to extract `visitReferencedFragmentsRecursively` that accept visitor function and if you return `false` it doesn't look inside current fragment.

Reply:
It only breaks `NoUnusedFragments` if you're using fragments with variable definitions. I *think* this is OK, as fragments with variable definitions are their own independent "unit": they have semantic meaning that is very different from normal fragments without variable definitions. An unused fragment that has variable definitions I see as being "stand-alone" in the same way a query is "stand-alone": you could reasonably figure out a way to persist these to the server and execute them, provided you started on the correct type.

```
fragment A($id: ID!) on Query {
  node(id: $id) {
    name
  }
}
```

is, and in my mind should be, identical to
```
query A($id: ID!) {
  node(id: $id) {
    name
  }
}
```

I'm very OK with breaking people who are actually using fragments with variable definitions.

I can see the appeal to extracting it out, but I'd much prefer taking an opinionated stance about what a fragment with variable definitions is. It will also make it easier to make an RFC in the spec for them.

Reply:
Ok, in that case, it's outside of scope for `graphql-js` and would make sense to discuss it in PR/RFC.

Just a quick note:
> and execute them, provided you started on the correct type

There is no one to one matching between root types and operation types so you can write:
```graphql
schema {
  query: Type
  mutation: Type
  subscription: Type
}
```

Reply:
Yeah after thinking this through I think it does make more sense to add a second function specifically for excluding fragments with variable definitions. I'll make that change.